### PR TITLE
Write deprecation logs to a data stream

### LINKED ISSUE #46106: Write deprecation logs to an index
The [Kibana Upgrade Assistant](https://www.elastic.co/guide/en/kibana/7.3/upgrade-assistant.html) helps users prepare to the next major version of Elasticsearch. The upgrade assistant works by introspecting various aspects of a cluster and its usage to surface deprecated functionality that is in use, and also to prepare indices that require re-indexing. The deprecations that are surfaced in the migration assistant are those that can be ascertained by introspecting the current state of the cluster. It can not however, catch on-going uses of deprecated functionality (e.g., APIs) that can not be ascertained by introspecting the current state of the cluster. The upgrade assistant would be even more useful for our users if we could assist users in understanding their use of such deprecated functionality.

Today we do surface such deprecated functionality via the deprecation logs. The upgrade assistant does not have an easy way to get its hands on the deprecation logs. 

The crux of this issue then is aimed at making it possible for the upgrade assistant to collect the deprecation logs from each running node. One way to do this is to write the deprecation logs to an index that the upgrade assistant could then read, along side the deprecations that it can already obtain.

It is likely that we want to consider the deprecation indices as system indices, and also manage them via ILM if ILM is available (e.g., use of a deprecated API no more recent than N months ago isn't relevant, it's likely that the user already migrated away from that API).

### PR DESCRIPTION
**Pull request by pugnascotia**
Backport of #58924.

Closes #46106. Introduce a mechanism for writing deprecation logs to a data stream as well as to disk.

Test by running `./gradlew run` and then:

```
curl -u elastic-admin:elastic-password http://localhost:9200/_flush/synced?pretty
curl -u elastic-admin:elastic-password http://localhost:9200/logs-deprecation-elasticsearch/_search?pretty
```

This implementation reworks deprecation logging to rely on log4j by introducing filters and appenders, and a custom log level. The existing `deprecation` X-Pack plugin simply configures an extra appender to capture messages to an index. A "component" class manages most of the interface between the cluster and the appender.

This approach can be extended for other types of logging, e.g. security by introducing further custom log levels.

### CODE REVIEW

File: x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/logging/DeprecationIndexingComponent.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+            .setFlushInterval(TimeValue.timeValueSeconds(5))
+            .build();
+    }
+
+    private static class DeprecationBulkListener implements BulkProcessor.Listener {
+        @Override
+        public void beforeBulk(long executionId, BulkRequest request) {}
+
+        @Override
+        public void afterBulk(long executionId, BulkRequest request, BulkResponse response) {}
```
Comment (martijnvg) [inline x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/logging/DeprecationIndexingComponent.java:139]: Can you add this to the body of the afterBulk() method to introspect if there are bulk item failures?

```
long items = request.numberOfActions();
if (logger.isTraceEnabled()) {
   logger.trace("indexed [{}] deprecation documents into [{}]", items,
   Arrays.stream(response.getItems())
      .map(BulkItemResponse::getIndex)
      .distinct()
      .collect(Collectors.joining(",")));
}

if (response.hasFailures()) {
   Map<String, String> failures = Arrays.stream(response.getItems())
      .filter(BulkItemResponse::isFailed)
      .collect(Collectors.toMap(BulkItemResponse::getId, BulkItemResponse::getFailureMessage));
   logger.error("failures: [{}]", failures);
}
```

**Comment by pugnascotia** [comment]
@pgomulka would you mind reviewing this backport, please? I was caught out by `ESJsonLayout` not generating ECS, and the simplest solution seemed to be forking it for `7.x`. I plan to forward-port the extra logging in `DeprecationIndexingComponent` as well.

**Review by pgomulka — APPROVED** [review]
I like the idea with forward fork
backporting ecs json changes to 7.x would break the json we introduced in 7.0 so would be annoying to users
LGTM when build is green

**Comment by pugnascotia** [comment]
Note that these changes were removed from `7.10` in #63942.

**Pull request merged**